### PR TITLE
Per Url zoom.default

### PR DESCRIPTION
**Pull request by toofar**
This change allows the `zoom.default` config setting to be configured using per-url settings. It consists of three main changes:

* Change all instances of `config.val.zoom.default` to config.instance.get('zoom.default', url=self._tab.url()) or similar,  see below for exceptions and caveats.
* Add a `AbstractTab.url_changed` listener to `AbstractZoom` to set the default zoom when navigating about.
* Modifying `AbstractZoom.set_factor()` and `WebEngineTab._restore_zoom()` to handle per url settings where the saved zoom value was the default zoom but no longer is after navigating.

Note the TODO I left in _on_url_changed: If someone sets a non-default zoom level for a tab then navigates to a page with a custom default zoom should we override the current tab zoom? I left it as ignoring the custom default zoom which is how everything currently works. This means that using url patterns with `zoom.default` is less useful for setting a custom zoom level for specific sites but then again, the name does say `zoom.default` not `zoom`.

Also there are two other small commits for related things that I ran into.

Oh I didn't add a test partly beause all the zoom tests are end2end ones.

## Caveats
Not all instances of `config.val.zoom.default` were changed because of issues with a circular dependancy at tab init time on the webengine backend. Getting the tab url depends on the tab widget being initialised and the function which sets the tab widget up (`AbstractTab._set_widget()` calls `zoom.set_default()`. This circular dependancy could probably be unpicked but I chose to just handle it by using the global zoom at init time and in `_init_neighborlist()` (which is called at init time as well as later) just fall back to the global setting on error. Or we could change `WebEngineTab.url()` to return `QUrl('')` if `_widget` hadn't been set yet.

The `AbstractTab.url_changed` signal can get fired when a navigation starts but before the new page has started to load, so the old one is still present. If the new page has a different default zoom the old page will change to that zoom at that point. I couldn't find a way around it (the other signals have their own problems) and in the usual case (fast internet, responsive sites) one wouldn't notice anyway.

`WebEngineTab._restore_zoom()` is a pain in the ass. I'm pretty sure I didn't break the fix but the `or_default` kwarg is a little weird. Other options:

* Make `AbstractZoom.factor()` take an optional `url` argument and it will return the zoom factor for that url instead of the current one. Have abstract tab pass in the url being navigated to. This works but
    feels like even more of a special-case than the current.
* Have `AbstractZoom.factor()` return some sentinal when the tab's default zoom hasn't been overriden. Returning something that isn't actually a float though would probably violate callers' expectations.
* Have `WebEngineTab._restore_zoom()` do more work to see if the tab is still using the default zoom and pass the appropriate default zoom to restore in that case rather than the saved one. Feels like an inversion of responsibilities.

---
This change is [Reviewable](https://reviewable.io/reviews/qutebrowser/qutebrowser/3782)

### CODE REVIEW

File: qutebrowser/browser/browsertab.py [hunk 1]
```diff
@@ -276,11 +279,24 @@ def __init__(self, tab, parent=None):
...
         #     cfg.changed.disconnect, self.init_neighborlist))
 
+    def _on_url_changed(self):
+        # NOTE: If a new page is requested and the request is slow to
+        # return data then this can be called while the "old" page is
+        # still displayed.
+        if self._default_zoom_changed:
+            # TODO: If someone has changed the zoom level for this tab
+            # then navigates to a tab with a custom default zoom should
+            # we change the tab zoom or expect people to press '='?
```
Comment (jgkamat) [inline qutebrowser/browser/browsertab.py]: I think that we can expect people to press `=`, I personally like the behavior here.

**Comment by The-Compiler** [comment]
I planned to take a look at this during my holidays before starting to learn for exams, but I didn't manage to do so (mainly because of a lot of issues with Qt 5.11...). I'll get back to it in September though!

**Comment by emdeeeks** [comment]
Are there any plans to fix the conflicts and merge? This is a feature I'd love to be able to use. :)

**Comment by toofar** [comment]
Yeah but no ETA. I haven't had much energy for open source work recently and I don't know if that will change. Also there are other PRs I would prefer to get in ahead of this. So if someone else wanted to take over I would encourage them to.

**Comment by toofar** [comment]
Closing this to make clear I don't intend to maintain it at this time. It's way too low on my priority list.

**Pull request closed without merging**